modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/fields/NestedFieldsSupplier.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.vulcan.fields;
+
+import com.liferay.petra.function.UnsafeFunction;
+import com.liferay.portal.kernel.util.ListUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Carlos Correa
+ */
+public class NestedFieldsSupplier<T> {
+
+	public List<String> getNestedFieldNames() {
+		if (_nestedFieldsContext == null) {
+			return new ArrayList<>();
+		}
+
+		return _nestedFieldsContext.getFieldNames();
+	}
+
+	public <T> T supply(
+			String fieldName,
+			UnsafeFunction<String, T, Exception> unsafeFunction)
+		throws Exception {
+
+		if (!_mustProcessNestedFields(_nestedFieldsContext)) {
+			return null;
+		}
+
+		List<String> fieldNames = _nestedFieldsContext.getFieldNames();
+
+		if (!fieldNames.contains(fieldName)) {
+			return null;
+		}
+
+		_nestedFieldsContext.incrementCurrentDepth();
+
+		try {
+			return unsafeFunction.apply(fieldName);
+		}
+		finally {
+			_nestedFieldsContext.decrementCurrentDepth();
+		}
+	}
+
+	public <T> Map<String, T> supply(
+			UnsafeFunction<String, T, Exception> unsafeFunction)
+		throws Exception {
+
+		if (!_mustProcessNestedFields(_nestedFieldsContext)) {
+			return null;
+		}
+
+		Map<String, T> nestedFields = new HashMap<>();
+
+		_nestedFieldsContext.incrementCurrentDepth();
+
+		for (String nestedFieldName : _nestedFieldsContext.getFieldNames()) {
+			T value = unsafeFunction.apply(nestedFieldName);
+
+			if (value != null) {
+				nestedFields.put(nestedFieldName, value);
+			}
+		}
+
+		_nestedFieldsContext.decrementCurrentDepth();
+
+		return nestedFields;
+	}
+
+	protected NestedFieldsSupplier(NestedFieldsContext nestedFieldsContext) {
+		_nestedFieldsContext =
```

Comment:
Picky comment, it is necessary the nestedFieldsContext of the constructor? Looking at this code seems like it is being ignored

Reply:
Ohh! You are right! It is not necessary, but after talking to Alex, we decided to include the context in the constructor as it is being built from the ContextThreadLocal, it is more clear where it comes from.

To be honest, I don't mind, whatever solution (including the constructor or not) it is ok to me.

Anyway, I am going to get the context from the attribute, if you consider we should change and get it from the context, it is ok to me

Reply:
Good catch!